### PR TITLE
CollectorUtils base implementation

### PR DESCRIPTION
**Pull request by kyzia551**

### CODE REVIEW

File: src/CollectorUtils.sol [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    if (swapperBalance < input.amount) {
+      input.amount = swapperBalance;
+    }
+
+    AaveSwapper(swapper).swap(
+      input.milkman,
+      input.priceChecker,
+      input.fromUnderlying,
+      input.toUnderlying,
+      input.fromUnderlyingPriceFeed,
```
Comment (sakulstra) [inline src/CollectorUtils.sol:171]: with recent proposal bugs perhaps good to fetch oracle decimals and ensure they are the same? (probably this should have better been on swapper itself, but as this is to late now, might make sense)

File: src/CollectorUtils.sol [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+  function __withdraw(
+    ICollector collector,
+    IOInput memory input,
+    address aTokenAddress
+  ) internal returns (uint256) {
+    if (input.amount == 0) {
+      revert InvalidZeroAmount();
+    }
+
+    if (input.amount == type(uint256).max) {
```
Comment (sakulstra) [inline src/CollectorUtils.sol]: not sure if this should be supported as the collector should never be emptied.

File: src/CollectorUtils.sol [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+      input.amount = IERC20(aTokenAddress).balanceOf(address(collector));
+    }
+    collector.transfer(aTokenAddress, address(this), input.amount);
+
+    // in case of imprecision during the aTokenTransfer withdraw a bit less
+    uint256 balanceAfterTransfer = IERC20(aTokenAddress).balanceOf(address(this));
+    input.amount = balanceAfterTransfer >= input.amount ? input.amount : balanceAfterTransfer;
+
+    // @dev withdrawal interfaces of v2 and v3 is the same, so we use any
+    IPool(input.pool).withdraw(input.underlying, input.amount, address(collector));
```
Comment (sakulstra) [inline src/CollectorUtils.sol]: not getting why this is needed - why not just withdraw `type(uint256).max`?

File: src/CollectorUtils.sol (hunk 1 above)
Comment (kyzia551) [reply src/CollectorUtils.sol:171]: Added [here](https://github.com/bgd-labs/aave-helpers/pull/326/files#diff-8405e99f7d86dbbd00abe684f6246d73267e7c0c6db7e9c2efb6e4de17119f78R148-R149)

File: src/CollectorUtils.sol (hunk 3 above)
Comment (kyzia551) [reply src/CollectorUtils.sol]: in case there will be some funds sent by mistake

File: src/CollectorUtils.sol [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    if (input.amount == 0) {
+      revert InvalidZeroAmount();
+    }
+
+    if (input.amount == type(uint256).max) {
+      input.amount = IERC20(aTokenAddress).balanceOf(address(collector));
+    }
+    collector.transfer(aTokenAddress, address(this), input.amount);
+    // @dev withdrawal interfaces of v2 and v3 is the same, so we use any
+    IPool(input.pool).withdraw(input.underlying, type(uint256).max, address(collector));
```
Comment (kyzia551) [inline src/CollectorUtils.sol]: Rephrased
cc: @sakulstra

**Review by sakulstra — APPROVED** [review]

**Pull request merged**